File: banana-algo-core/02-搜索算法/dfs_bfs.cpp
/**
 * DFS和BFS实现
 */

#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_set>
#include <unordered_map>

using namespace std;

// DFS递归
void dfsRecursive(const unordered_map<int, vector<int>>& graph, 
                  int node, unordered_set<int>& visited) {
    visited.insert(node);
    cout << node << " ";
    
    if (graph.count(node)) {
        for (int neighbor : graph.at(node)) {
            if (!visited.count(neighbor)) {
                dfsRecursive(graph, neighbor, visited);
            }
        }
    }
}

// DFS迭代（栈）
vector<int> dfsIterative(const unordered_map<int, vector<int>>& graph, int start) {
    unordered_set<int> visited;
    stack<int> st;
    vector<int> result;
    
    st.push(start);
    
    while (!st.empty()) {
        int node = st.top();
        st.pop();
        
        if (!visited.count(node)) {
            visited.insert(node);
            result.push_back(node);
            
            if (graph.count(node)) {
                auto neighbors = graph.at(node);
                for (auto it = neighbors.rbegin(); it != neighbors.rend(); ++it) {
                    if (!visited.count(*it)) {
                        st.push(*it);
                    }
                }
            }
        }
    }
    
    return result;
}

// BFS（队列）
vector<int> bfs(const unordered_map<int, vector<int>>& graph, int start) {
    unordered_set<int> visited;
    queue<int> q;
    vector<int> result;
    
    visited.insert(start);
    q.push(start);
    
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        result.push_back(node);
        
        if (graph.count(node)) {
            for (int neighbor : graph.at(node)) {
                if (!visited.count(neighbor)) {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }
    }
    
    return result;
}

// 岛屿数量
void dfsIsland(vector<vector<char>>& grid, int i, int j) {
    int m = grid.size();
    int n = grid[0].size();
    
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') {
        return;
    }
    
    grid[i][j] = '0';
    
    dfsIsland(grid, i + 1, j);
    dfsIsland(grid, i - 1, j);
    dfsIsland(grid, i, j + 1);
    dfsIsland(grid, i, j - 1);
}

int numIslands(vector<vector<char>>& grid) {
    if (grid.empty()) return 0;
    
    int count = 0;
    int m = grid.size();
    int n = grid[0].size();
    
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == '1') {
                count++;
                dfsIsland(grid, i, j);
            }
        }
    }
    
    return count;
}

int main() {
    cout << "=== DFS和BFS演示 ===" << endl << endl;
    
    // 构建图
    unordered_map<int, vector<int>> graph;
    graph[0] = {1, 2};
    graph[1] = {0, 3, 4};
    graph[2] = {0, 5};
    graph[3] = {1};
    graph[4] = {1, 5};
    graph[5] = {2, 4};
    
    cout << "图结构:" << endl;
    for (const auto& pair : graph) {
        cout << "  " << pair.first << ": [";
        for (size_t i = 0; i < pair.second.size(); i++) {
            cout << pair.second[i];
            if (i < pair.second.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }
    cout << endl;
    
    // DFS
    cout << "DFS遍历（递归）: ";
    unordered_set<int> visited;
    dfsRecursive(graph, 0, visited);
    cout << endl << endl;
    
    cout << "DFS遍历（迭代）: ";
    vector<int> dfs_result = dfsIterative(graph, 0);
    for (int node : dfs_result) {
        cout << node << " ";
    }
    cout << endl << endl;
    
    // BFS
    cout << "BFS遍历: ";
    vector<int> bfs_result = bfs(graph, 0);
    for (int node : bfs_result) {
        cout << node << " ";
    }
    cout << endl << endl;
    
    // 岛屿数量
    cout << "岛屿数量:" << endl;
    vector<vector<char>> grid = {
        {'1', '1', '0', '0', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'}
    };
    
    for (const auto& row : grid) {
        cout << "  ";
        for (char cell : row) {
            cout << cell << " ";
        }
        cout << endl;
    }
    cout << "岛屿数量: " << numIslands(grid) << endl;
    
    return 0;
}

File: web-sites-hub/banana-algo-core-docs/02-search-algorithms/dfs_bfs.py
"""
DFS和BFS实现
"""

from collections import deque

# ========== DFS深度优先 ==========

def dfs_recursive(graph, start, visited=None):
    """DFS递归实现"""
    if visited is None:
        visited = set()
    
    visited.add(start)
    print(start, end=' ')
    
    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs_recursive(graph, neighbor, visited)
    
    return visited


def dfs_iterative(graph, start):
    """DFS迭代实现（栈）"""
    visited = set()
    stack = [start]
    result = []
    
    while stack:
        node = stack.pop()
        if node not in visited:
            visited.add(node)
            result.append(node)
            
            # 逆序添加邻居（保证顺序）
            for neighbor in reversed(graph[node]):
                if neighbor not in visited:
                    stack.append(neighbor)
    
    return result


# ========== BFS广度优先 ==========

def bfs(graph, start):
    """BFS实现（队列）"""
    visited = {start}
    queue = deque([start])
    result = []
    
    while queue:
        node = queue.popleft()
        result.append(node)
        
        for neighbor in graph[node]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)
    
    return result


def bfs_shortest_path(graph, start, end):
    """BFS求最短路径"""
    if start == end:
        return [start]
    
    visited = {start}
    queue = deque([(start, [start])])
    
    while queue:
        node, path = queue.popleft()
        
        for neighbor in graph[node]:
            if neighbor not in visited:
                new_path = path + [neighbor]
                if neighbor == end:
                    return new_path
                visited.add(neighbor)
                queue.append((neighbor, new_path))
    
    return None


def bfs_levels(graph, start):
    """BFS分层遍历"""
    visited = {start}
    queue = deque([start])
    levels = []
    
    while queue:
        level_size = len(queue)
        current_level = []
        
        for _ in range(level_size):
            node = queue.popleft()
            current_level.append(node)
            
            for neighbor in graph[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        
        levels.append(current_level)
    
    return levels


# ========== 应用：岛屿数量 ==========

def num_islands(grid):
    """岛屿数量（DFS）"""
    if not grid:
        return 0
    
    m, n = len(grid), len(grid[0])
    count = 0
    
    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == '0':
            return
        
        grid[i][j] = '0'  # 标记已访问
        
        # 四个方向
        dfs(i + 1, j)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i, j - 1)
    
    for i in range(m):
        for j in range(n):
            if grid[i][j] == '1':
                count += 1
                dfs(i, j)
    
    return count


# ========== 应用：课程表（拓扑排序） ==========

def can_finish(num_courses, prerequisites):
    """判断能否完成所有课程（检测环）"""
    graph = [[] for _ in range(num_courses)]
    for course, prereq in prerequisites:
        graph[prereq].append(course)
    
    visited = [0] * num_courses  # 0:未访问 1:访问中 2:已完成
    
    def has_cycle(node):
        if visited[node] == 1:
            return True  # 发现环
        if visited[node] == 2:
            return False
        
        visited[node] = 1
        for neighbor in graph[node]:
            if has_cycle(neighbor):
                return True
        visited[node] = 2
        return False
    
    for i in range(num_courses):
        if visited[i] == 0:
            if has_cycle(i):
                return False
    
    return True


def demo():
    """演示DFS和BFS"""
    print("=== DFS和BFS演示 ===\n")
    
    # 构建图
    graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D', 'E'],
        'C': ['A', 'F'],
        'D': ['B'],
        'E': ['B', 'F'],
        'F': ['C', 'E']
    }
    
    print("图结构:")
    for node, neighbors in graph.items():
        print(f"  {node}: {neighbors}")
    print()
    
    print("DFS遍历（递归）: ", end='')
    dfs_recursive(graph, 'A')
    print("\n")
    
    print("DFS遍历（迭代）:", dfs_iterative(graph, 'A'))
    print("BFS遍历:", bfs(graph, 'A'))
    print()
    
    # 最短路径
    path = bfs_shortest_path(graph, 'A', 'F')
    print(f"从A到F的最短路径: {' -> '.join(path) if path else 'None'}")
    print()
    
    # 分层遍历
    print("BFS分层遍历:")
    levels = bfs_levels(graph, 'A')
    for i, level in enumerate(levels):
        print(f"  Level {i}: {level}")
    print()
    
    # 岛屿数量
    grid = [
        ['1', '1', '0', '0', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '1', '0', '0'],
        ['0', '0', '0', '1', '1']
    ]
    print("岛屿数量:")
    for row in grid:
        print(f"  {row}")
    # 注意：会修改grid，这里重新创建
    grid_copy = [row[:] for row in grid]
    print(f"岛屿数量: {num_islands(grid_copy)}")


if __name__ == '__main__':
    demo()

